Feat: Home UI 구현

Jeon0976 opened the pull request:
## Motivation ⍰

- 홈 화면 UI 구현 
- 1. NagazaBaseViewController: adjustLayoutAfterRendering 한 번만 호출되도록 로직 추가 
- 2. NagazaTabBarController: 기종마다 바텀 safeAreaInsets 값에 맞춰서 inset값 반영되도록 변경 

<br>

## Key Changes 🔑

- 아이콘을 제외한 홈 화면 기본 UI 구현했습니다. 참고 부탁드릴께요. 
- 1. 내용 코드 입니다 .
``` swift 
    private var afterViewDidLoad = true

    override func viewDidLayoutSubviews() {
        super.viewDidLayoutSubviews()
        
        if afterViewDidLoad {
            afterViewDidLoad.toggle()
            adjustLayoutAfterRendering()
        }
    }
``` 
- 2. 내용 코드 입니다. 
- 2-1. `tabBarHeight` 전역 상수로 만들었습니다. home, map, review, myPage ViewController에서 Bottom inset을 정할때 superView를 기준으로 할때 해당 inset값을 추가해야 합니다. 
    - 예시)  `$0.bottom.equalTo(view.safeAreaLayoutGuide.snp.bottom).inset(tabBarHeight)`
``` swift 
let tabBarHeight: CGFloat = 56

final class NagazaTabBarController: NagazaBaseViewController {
...
    override func adjustLayoutAfterRendering() {
        updateTabBarHeight()
        setupButtons(with: types)
    }

   private func updateTabBarHeight() {
        tabBarView.snp.makeConstraints {
            $0.height.equalTo(tabBarHeight + view.safeAreaInsets.bottom)
        }
    }
    
    private func setupButtons(with types: [TabBarType]) {
        let buttonWidth = view.bounds.width / CGFloat(types.count)
        
        for (index, type) in types.enumerated() {
            let button = TabBarButton(tabBarType: type)
            button.tag = index
            button.addTarget(self, action: #selector(tabBarButtonTapped(_:)), for: .touchUpInside)
            
            tabBarView.addSubview(button)
            buttons.append(button)
            
            button.snp.makeConstraints {
                $0.leading.equalToSuperview().offset(buttonWidth * CGFloat(index))
                $0.width.equalTo(buttonWidth)
                
                let viewBottom = view.safeAreaInsets.bottom
                var inset: CGFloat = 0
                if viewBottom > 0 {
                    inset = viewBottom / 2
                }
                
                $0.height.equalTo(tabBarHeight + inset)
            }
        }
    }
...
}
```
<br>

## To Reviewers 🙏🏻

- 

<br>

## Linked Issue 🔗

-

Jeon0976 committed:
Feat: Home UI 구현

Jeon0976 committed:
Refector: CollectionVC, adjustLayoutAfterRendering 기능 한 번만 되도록 수정

Jeon0976 committed:
Feat: NagazaBaseTableViewController 추가

Jeon0976 committed:
Fix: HomeThemesViewController 오타 수정

Minny27 commented on Nagaza/Sources/Presentation/Feature/TabBar/Main/View/HomeViewSubView/HomeThemesViewController.swift:
```diff
@@ -14,10 +14,14 @@ final class HomeThemesViewController: NagazaBaseViewController {
         let label = UILabel()
-        
+
+        label.translatesAutoresizingMaskIntoConstraints = false
```
snapKit으로 autolayout 잡으면 내부적으로 설정해줘서 translatesAutoresizingMaskIntoConstaints 설정 안 해도 돼요

Minny27 commented on Nagaza/Sources/Presentation/Base/NagazaBaseCollectionViewController.swift:
```diff
@@ -11,7 +11,9 @@ import RxSwift
     var disposeBag = DisposeBag()
-    
+
+    private var afterViewDidLoad = true
```
이 변수는 어떤 의미인지 잘 모르겠네요

Minny27 replied:
### 수정 사항
1. tab bar의 높이는 56으로 고정시키고 아이템의 상단 패딩을 5설정
- 버튼 아이템이 수직 센터로 맞춰져 있는데 디자인 가이드를 맞추기 위해서 패딩을 설정
- 그리고 노치가 있는 폰은 하단의 남는 부분을 채우기 위해서 BottomInsetView를 따로 만듦.
- 이렇게 해야 바텀 inset 크기에 상관없이 동일한 크기와 위치의 탭 바를 만들 수 있음

|수정 전|수정 후|
|:---:|:---:|
|<img src="https://github.com/NAGAZA-Team/NAGAZA-iOS/assets/68800789/ff9dc051-e89a-4a30-a6e1-95a05068d79e" width=80%>|<img src="https://github.com/NAGAZA-Team/NAGAZA-iOS/assets/68800789/030881ab-5ea5-46ba-b683-da035b271734" width=80%>|

2. 탭 바 스크롤 뷰 하단을 탭 바의 탑에 맞추기
- 먼저 tabBarHeight을 NagazaSize enum에 static 상수로 정의
- NagazaBaseScrollView를 정의해서 contentInset으로 bottom을 탭 바 높이만큼 설정
- 이렇게 하면 탭 바 높이만큼 더 보이게 할 수 있음
코드는 다음과 같음
```swift
// NagazaBaseScrollView
class NagazaBaseScrollView: UIScrollView {
    init() {
        super.init(frame: .zero)
        self.contentInset = UIEdgeInsets(top: 0,
                                         left: 0,
                                         bottom: NagazaSize.tabBarHeight,
                                         right: 0)
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

// HomeViewController
private lazy var scrollView = NagazaBaseScrollView()

```

Jeon0976 replied:
> 이 변수는 어떤 의미인지 잘 모르겠네요

viewdidlayoutSubviews 메서드가 어떤 특정 행위를 할때마다 호출되는 함수여서 viewdidload후에만 활동되도록 하기 위해 만든 변수입니다.

Minny27 committed:
[#5]Refactor: 탭 바 디자인 가이드와 일치시키기

- 탭 바 제약조건 한 번에 처리
- bottom inset이 디바이스마다 다르기 떄문에
- Tab bar의 bottom을 safeArea에 맞추고 탭 바 높이로 설정(bottom inset 사용 x)
- 탭 바 item의 높이도 탭 바 높이로 설정
- 그런데 수직 중앙 정렬이 되어, 디자인 가이드에 맞게 top padding 값을 추가

Minny27 committed:
[#5]Refactor: 스크롤 뷰 바텀을 탭 바 상단으로 맞추기

- 먼저 탭 바 높이는 NagazaSize enum에 static 상수로 선언
- 스크롤 뷰를 커스텀해서 하단 inset을 tab bar height만큼 설정
- 이렇게 하면 scroll의 bottom을 safe area bottom으로 맞추면 탭 바 길이만큼 더 보여줌
- 탭 바에 아이템이 기본적으로 수직 센터 정렬로 되어 있어, top padding을 5만큼 설정해야 함.
- 그래서 탭 바의 크기는 56으로 고정으로 가야하기 떄문에 bottom inset으로 높이를 설정하지 않음.
- 대신에 bottomInsetView를 만들어서 노치있는 부분을 채우는 방식으로 구현

Minny27 reviewed (approved):
고생하셨습니다.
